### ISSUE TITLE
Cast bug

### ISSUE THREAD

**Issue by Fallenyasha** [ClickHouse/ClickHouse#9810]
Received exception: DB::Exception: Second argument to CAST must be a constant string describing type (version 19.17.2.4 (official build))
`select cast(1 as String)`
`from (select 1 as iid) as t1`
`join (select '1' as sid) as t2 on t2.sid = cast(t1.iid as String)`

On version 19.4.5.1 this query works fine.

Also if I change some parts in this query it will be work without exceptions, for example:
`select cast(1 as Date)` -- or another type except String
`from (select 1 as iid) as t1`
`join (select '1' as sid) as t2 on t2.sid = cast(t1.iid as String)`

`select cast(1 as String)`
`from (select 1 as iid) as t1`

`select cast(1 as String)` -- or another type
`from (select 1 as iid) as t1`
`join (select '1' as sid) as t2 on cast(t2.sid as UInt8) = t1.iid`

**Comment by qoega** [comment]
reproduces in master, 20.3

**Comment by 4ertus2** [comment]
Related #7798

**Comment by akuzm** [comment]
Can be reproduced even simpler:

```
:) select cast(2 as String), * from (select 2 "'String'")

Received exception from server (version 20.4.1):
Code: 43. DB::Exception: Received from localhost:9000. DB::Exception: Second argument to CAST must be a constant string describing type. Stack trace:
```

Technical note:
This is a general issue where auto-generated column names collide with other names. In the original case, column (1) is a constant column `"'String'"` with value `String`, generated as a second parameter for the function `cast(1, 'String')`. Column (2) is identical, but generated for `cast(t1.iid, 'String')`. Column (2) is below the join, so it gets materialized when JOIN expression action is applied. After that, `cast(1, 'String')` tries to use the same column and finds that it is not constant, hence the error.

**Comment by alexey-milovidov** [comment]
```
SELECT concat('xyz', 'abc'), * FROM (SELECT 2 AS `'xyz'`)
```

**Comment by alexey-milovidov** [comment]
@akuzm @4ertus2 To solve this issue, we have to assign unique names only to constant expressions that will be materialized and altered during JOINs.

e.g. LEFT JOINed constants that may become NULL.

**Comment by alexey-milovidov** [comment]
The most illustrating example for the root cause of this issue:

```
SELECT 
    t1.*, 
    t2.*, 
    'world', 
    isConstant('world')
FROM 
(
    SELECT 
        arrayJoin([1, 2]) AS k, 
        'hello'
) AS t1
LEFT JOIN 
(
    SELECT 
        arrayJoin([1, 3]) AS k, 
        'world'
) AS t2 ON t1.k = t2.k

┌─k─┬─'hello'─┬─t2.k─┬─'world'─┬─'world'─┬─isConstant('world')─┐
│ 1 │ hello   │    1 │ world   │ world   │                   0 │
│ 2 │ hello   │    0 │         │         │                   0 │
└───┴─────────┴──────┴─────────┴─────────┴─────────────────────┘
```

**Comment by alexey-milovidov** [comment]
https://github.com/ClickHouse/ClickHouse/issues/9810#issuecomment-604995430
https://github.com/ClickHouse/ClickHouse/issues/9810#issuecomment-605214756
^ these two examples are not very relevant because the user intentionally hacked the query analysis with unusual aliases. They not necessary to be fixed.

**Comment by alexey-milovidov** [comment]
What about the following logic:

If it's JOIN action and we have a column with constant expression and without an alias, then assign a unique name.

**Comment by alexey-milovidov** [comment]
It may be simply done by assigning qualified names.

**Comment by 4ertus2** [comment]
Draft for fix constants from left table #9950
